File: go/300.500/432.AllOOneDataStructure.go
package main

import (
	"container/list"
	"fmt"
)

type node struct {
	keys  map[string]struct{}
	count int
}

type AllOne struct {
	lst   *list.List
	nodes map[string]*list.Element
}

func Constructor() AllOne {
	return AllOne{
		lst:   list.New(),
		nodes: map[string]*list.Element{},
	}
}

func (this *AllOne) Inc(key string) {
	if cur := this.nodes[key]; cur != nil {
		curNode := cur.Value.(node)
		if nxt := cur.Next(); nxt == nil || nxt.Value.(node).count > curNode.count+1 {
			this.nodes[key] = this.lst.InsertAfter(node{map[string]struct{}{key: {}}, curNode.count + 1}, cur)
		} else {
			nxt.Value.(node).keys[key] = struct{}{}
			this.nodes[key] = nxt
		}
		delete(curNode.keys, key)
		if len(curNode.keys) == 0 {
			this.lst.Remove(cur)
		}
	} else {
		if this.lst.Front() == nil || this.lst.Front().Value.(node).count > 1 {
			this.nodes[key] = this.lst.PushFront(node{map[string]struct{}{key: {}}, 1})
		} else {
			this.lst.Front().Value.(node).keys[key] = struct{}{}
			this.nodes[key] = this.lst.Front()
		}
	}
}

func (this *AllOne) Dec(key string) {
	cur := this.nodes[key]
	curNode := cur.Value.(node)
	if curNode.count > 1 {
		if pre := cur.Prev(); pre == nil || pre.Value.(node).count < curNode.count-1 {
			this.nodes[key] = this.lst.InsertBefore(node{map[string]struct{}{key: {}}, curNode.count - 1}, cur)
		} else {
			pre.Value.(node).keys[key] = struct{}{}
			this.nodes[key] = pre
		}
	} else {
		delete(this.nodes, key)
	}
	delete(curNode.keys, key)
	if len(curNode.keys) == 0 {
		this.lst.Remove(cur)
	}
}

func (this *AllOne) GetMaxKey() string {
	if b := this.lst.Back(); b != nil {
		for key := range b.Value.(node).keys {
			return key
		}
	}
	return ""
}

func (this *AllOne) GetMinKey() string {
	if b := this.lst.Front(); b != nil {
		for key := range b.Value.(node).keys {
			return key
		}
	}
	return ""
}

func main() {
	cls := Constructor()
	cls.Inc("a")
	cls.Inc("b")
	cls.Inc("b")
	cls.Inc("c")
	cls.Inc("c")
	cls.Inc("c")
	cls.Dec("b")
	cls.Dec("b")
	fmt.Println("Min:", cls.GetMinKey())
	cls.Dec("a")
	fmt.Println("Max:", cls.GetMaxKey())
	fmt.Println("Min:", cls.GetMinKey())
}

File: 300.500/432.AllOOneDataStructure.cpp
#include <iostream>
#include <string>
#include <list>
#include <unordered_set>
#include <unordered_map>
#include <map>
using namespace std;

class AllOne {
public:
    AllOne() {

    }
    
    void inc(string key) {
        if (nodes.count(key) > 0) {
            //已经存在的key
            auto cur = nodes[key];
            auto nxt = next(cur);
            if (nxt == lst.end() || nxt->second > cur->second + 1) {
                //需要添加新的节点
                unordered_set<string> s{ key };
                nodes[key] = lst.emplace(nxt, s, cur->second + 1);
            }
            else {
                //在老节点添加元素
                nxt->first.emplace(key);
                nodes[key] = nxt;
            }
            //删除原节点
            cur->first.erase(key);
            if (cur->first.empty()) {
                lst.erase(cur);
            }
        }
        else {
            if (lst.empty() || lst.begin()->second > 1) {
                unordered_set<string> s{ key };
                lst.emplace_front(s, 1);
            }
            else {
                lst.begin()->first.emplace(key);
            }
            nodes[key] = lst.begin();
        }
    }
    
    void dec(string key) {
        auto cur = nodes[key];
        if (cur->second == 1) {
            nodes.erase(key);
        }
        else {
            auto pre = prev(cur);
            if (cur == lst.begin() || pre->second < cur->second - 1) {
                unordered_set<string> s{ key };
                nodes[key] = lst.emplace(cur, s, cur->second - 1);
            }
            else {
                pre->first.emplace(key);
                nodes[key] = pre;
            }
        }
        cur->first.erase(key);
        if (cur->first.empty()) {
            lst.erase(cur);
        }
    }
    
    string getMaxKey() {
        return lst.empty() ? "" : *lst.rbegin()->first.begin();
    }
    
    string getMinKey() {
        return lst.empty() ? "" : *lst.begin()->first.begin();
    }
private:
    list<pair<unordered_set<string>, int>> lst;
    unordered_map<string, list<pair<unordered_set<string>, int>>::iterator> nodes;
};

class AllOne1 {
public:
    AllOne1() {

    }
    
    void inc(string key) {
        int cnt = nodes[key];
        nodes[key] = cnt+1;
        if (cnt > 0) {
            dp[cnt].erase(key);
            if (dp[cnt].empty()) {
                dp.erase(cnt);
            }
        }
        dp[cnt+1].emplace(key);
    }
    
    void dec(string key) {
        int cnt = nodes[key];
        if (cnt == 1) {
            nodes.erase(key);
        }
        else {
            nodes[key] = cnt - 1;
            dp[cnt - 1].emplace(key);
        }
        dp[cnt].erase(key);
        if (dp[cnt].empty()) {
            dp.erase(cnt);
        }
    }
    
    string getMaxKey() {
        return dp.empty() ? "" : *dp.rbegin()->second.begin();
    }
    
    string getMinKey() {
        return dp.empty() ? "" : *dp.begin()->second.begin();
    }
private:
    unordered_map<string, int> nodes;
    map<int, unordered_set<string>> dp;
};

int main(int argc, char const* argv[]) {
    AllOne1 cls;
    cls.inc("a");
    cls.inc("b");
    cls.inc("b");
    cls.inc("c");
    cls.inc("c");
    cls.inc("c");
    cls.dec("b");
    cls.dec("b");
    cout << "Min:" << cls.getMinKey().c_str() << endl;
    cls.dec("a");
    cout << "Max:" << cls.getMaxKey().c_str() << endl;
    cout << "Min:" << cls.getMinKey().c_str() << endl;
    // cls.inc("a");
    // cls.inc("b");
    // cls.inc("b");
    // cls.inc("b");
    // cls.inc("b");
    // cls.dec("b");
    // cls.dec("b");
    // cout << "Max:" << cls.getMaxKey().c_str() << endl;
    // cout << "Min:" << cls.getMinKey().c_str() << endl;
    // cls.inc("hello");
    // cls.inc("hello");
    // cout << "Max:" << cls.getMaxKey().c_str() << endl;
    // cout << "Min:" << cls.getMinKey().c_str() << endl;
    // cls.inc("leet");
    // cout << "Max:" << cls.getMaxKey().c_str() << endl;
    // cout << "Min:" << cls.getMinKey().c_str() << endl;
    return 0;
}
